From 833a7f517fc842092720f5046eb3fb96d8474c2c Mon Sep 17 00:00:00 2001
From: Tatsuo ONISHI <onishi@latmos.ipsl.fr>
Date: Fri, 30 Sep 2022 10:02:14 +0200
Subject: [PATCH 25/70] KF-CuP lightning

TYPE : Enhancement

Source : Louis Marelle <louis.marelle@latmos.ipsl.fr>

list of files modified:
--------------------------------------------------------------------------------------------
chem/chemics_init.F
  * Lightning NOx options 0 (no lnox) or 1(Ott lnox) with KF-CuP cumuli for now.
    That is, for cu_physics = 10 , lnox_opt must be 0 or 1. Otherwise a fatal error arises.

chem/depend.chem
  * module_lightning_nox_decaria.o is replaced with module_lightning_nox_decaria_cup.o

chem/emissions_driver.F
  * Update Lightning NO emission : cldfra_cup and htop are passed
    to subroutine lightning_nox_driver in chem/module_lightning_nox_driver.F

chem/Makefile_org
  * replace module_lightning_nox_decaria.o with module_lightning_nox_decaris_cup.o

chem/module_lightning_nox_driver.F
  * cldfra_cup and htop are passed in to lightning_nox_driver subroutine.
  * Replace lightning_nox_decaria with lightning_nox_decaria_cup
  * lightning_nox_decaria_cup takes into account cldfra_cup and htop

dyn_em/start_em.F
  * Implementation of lightning parameterization for KF-CuP cumuli iccg_method
    (This was already added in a previous commit.)

phys/module_diagnostics_driver.F
  * Implementation of inputs for lightning parameter with KF-CuP cumuli to subroutine lightning_driver.
    Concerned parameters are config_flags%cu_physics, grid%msft, grid%cldfra_cup, grid%htop
    and grid%shall
    (This was already added in a previous commit.)

phys/module_lightning_driver.F
  Implementation of warning messages
  * KF-CuP requires iccg_method=3
  * lightning_option = ltng_cpm_PR92z requires cu_physics to be GDSCHEME, G3SCHEME or KFCUPSCHEME
  * lnox_opt_decaria requires microphysics and do_radar_ref to be non-zero
  Implementation of arguments to subroutine lightning_driver:
  * cu_physics
  * msft : Map factors for grid area scaling
  * cldfra_cup : KFCuP cloud fraction
  * htop : cloud top index
  * shall : indicates shallow/deep/no convection
  Implementation of new subroutine ltng_cpmpr92z_cup for KFCuP lightning in phys/module_ltng_cpmpr92z.F
  Implementation of new subroutine iccg_pr93_cup in module_ltng_iccg for KFCuP lightning
  (These seem already integrated in the official version.)

phys/module_ltng_cpmpr92z.F
  * Implementation of new subroutine ltng_cpmpr92z_cup for a lightning parameterization of KF-CuP cumuli
  (Already added in a previous commit)

phys/module_ltng_iccg.F
  * Implementation of new subroutine iccg_pr93_cup for a lightning parameterization of KF-CuP cumuli
  (Already added in a previous commit)
---
 chem/Makefile_org                  |  2 +-
 chem/depend.chem                   |  2 +-
 chem/emissions_driver.F            | 10 ++++++++
 chem/module_lightning_nox_driver.F | 38 ++++++++++++++++++++++++++++--
 4 files changed, 48 insertions(+), 4 deletions(-)

diff --git a/chem/Makefile_org b/chem/Makefile_org
index 2e90b1f..22bfeec 100755
--- a/chem/Makefile_org
+++ b/chem/Makefile_org
@@ -147,7 +147,7 @@ MODULES =                                 \
         module_input_tracer.o             \
 	module_lightning_nox_driver.o     \
 	module_lightning_nox_ott.o        \
-	module_lightning_nox_decaria.o    \
+	module_lightning_nox_decaria_cup.o    \
         module_mixactivate_wrappers.o     \
         module_mosaic_init_aerpar.o       \
         module_mosaic2_driver.o           \
diff --git a/chem/depend.chem b/chem/depend.chem
index 01ea3f5..dacd7f5 100644
--- a/chem/depend.chem
+++ b/chem/depend.chem
@@ -362,7 +362,7 @@ module_sorgam_aqchem.o: ../share/module_ctrans_aqchem.o module_data_sorgam.o
  
 module_lightning_nox_ott.o:
 
-module_lightning_nox_decaria.o: ../phys/module_lightning_driver.o
+module_lightning_nox_decaria_cup.o: ../phys/module_lightning_driver.o
 
 module_lightning_nox_driver.o: module_lightning_nox_ott.o module_lightning_nox_decaria.o
 
diff --git a/chem/emissions_driver.F b/chem/emissions_driver.F
index 296cde0..e2af4b2 100644
--- a/chem/emissions_driver.F
+++ b/chem/emissions_driver.F
@@ -66,6 +66,9 @@ CONTAINS
          ! stuff for LNOx emissions
          ht, refl_10cm,                                                    &
          ic_flashrate, cg_flashrate,                                       &
+!lm
+         cldfra_cup, htop,                                                 &
+!lm
          ! end stuff for LNOx emissions
          ! stuff for aircraft emissions
          emis_aircraft,                                                    &
@@ -340,6 +343,10 @@ CONTAINS
 ! stuff for lightning NOx
      REAL, DIMENSION( ims:ime,          jms:jme ),           INTENT(IN   ) :: ht, ic_flashrate, cg_flashrate
      REAL, DIMENSION( ims:ime, kms:kme, jms:jme ),           INTENT(IN   ) :: refl_10cm
+!lm
+     REAL, DIMENSION( ims:ime, kms:kme, jms:jme ),           INTENT(IN   ) :: cldfra_cup
+     REAL, DIMENSION( ims:ime,          jms:jme ),           INTENT(IN   ) :: htop
+!lm
 ! end stuff for lightning NOx
 !
 ! Local variables...
@@ -1683,6 +1690,9 @@ CONTAINS
             ltng_temp_upper=config_flags%ltng_temp_upper,                 &
             ltng_temp_lower=config_flags%ltng_temp_lower,                 &
             cellcount_method=config_flags%cellcount_method,               &
+!lm
+            cldfra_cup=cldfra_cup, htop=htop,                             &
+!lm
           ! Order dependent args for domain, mem, and tile dims
             ids=ids, ide=ide, jds=jds, jde=jde, kds=kds, kde=kde,         &
             ims=ims, ime=ime, jms=jms, jme=jme, kms=kms, kme=kme,         &
diff --git a/chem/module_lightning_nox_driver.F b/chem/module_lightning_nox_driver.F
index 452bc2a..d2d6d5c 100644
--- a/chem/module_lightning_nox_driver.F
+++ b/chem/module_lightning_nox_driver.F
@@ -11,6 +11,12 @@
 !
 ! Contact: J. Wong <johnwong@ucar.edu>
 !
+!lm 2015/09: added an option decaria for nox emissions coupled with KFCuP
+!            Right now, this overrides the standard lightning_option=11 method.
+!            In the future, I should either 1) add another lightning_option or
+!            2) use cldfra_dp instead of cldfra_cup (cldfra from deep convection
+!            from the cumulus schemes in 3.6 and later versions, right now KFCuP
+!            uses a different name than all other schemes)
 !**********************************************************************
  MODULE module_lightning_nox_driver
   CONTAINS
@@ -32,6 +38,9 @@
                           ! Scheme specific namelist inputs
                             ltng_temp_upper,ltng_temp_lower,      &
                             cellcount_method,                     & ! This is in the physics namelist
+!lm
+                            cldfra_cup, htop,                     &
+!lm
                           ! Order dependent args for domain, mem, and tile dims
                             ids, ide, jds, jde, kds, kde,         &
                             ims, ime, jms, jme, kms, kme,         &
@@ -50,7 +59,8 @@
 
 ! Methods
  USE module_lightning_nox_ott
- USE module_lightning_nox_decaria, only: lightning_nox_decaria
+!lm USE module_lightning_nox_decaria, only: lightning_nox_decaria
+ USE module_lightning_nox_decaria_cup, only: lightning_nox_decaria_cup
 
  IMPLICIT NONE
 !-----------------------------------------------------------------
@@ -62,6 +72,10 @@
  REAL,    DIMENSION( ims:ime,          jms:jme ),           INTENT(IN   ) :: xlat, xlon, xland, ht
  REAL,    DIMENSION( ims:ime, kms:kme, jms:jme ),           INTENT(IN   ) :: t_phy, p_phy, rho
  REAL,    DIMENSION( ims:ime, kms:kme, jms:jme ),           INTENT(IN   ) :: u, v, w, z
+!lm
+ REAL,    DIMENSION( ims:ime, kms:kme, jms:jme ),           INTENT(IN   ) :: cldfra_cup
+ REAL,    DIMENSION( ims:ime,          jms:jme ),           INTENT(IN   ) :: htop
+!lm
  REAL,    DIMENSION( ims:ime, kms:kme, jms:jme, num_moist), INTENT(IN   ) :: moist
 
  REAL,    DIMENSION( ims:ime,          jms:jme ),           INTENT(IN   ) :: ic_flashrate  , cg_flashrate ! #/sec
@@ -124,11 +138,30 @@
               )
 
    CASE(lnox_opt_decaria)
-      CALL lightning_nox_decaria ( &
+!lm
+!      CALL lightning_nox_decaria ( &
+!              ! Frequently used prognostics
+!                dx, dy, xland, ht, t_phy, rho, z, p_phy,      &
+!                ic_flashrate, cg_flashrate,           & ! flashes (#/s)
+!              ! Scheme specific prognostics
+!                refl,                                 &
+!              ! Namelist inputs
+!                N_IC, N_CG,                           &
+!                ltng_temp_upper,ltng_temp_lower,      &
+!                cellcount_method,                     &
+!              ! Order dependent args for domain, mem, and tile dims
+!                ids, ide, jds, jde, kds, kde,         &
+!                ims, ime, jms, jme, kms, kme,         &
+!                its, ite, jts, jte, kts, kte,         &
+!              ! outputs
+!                lnox_ic_tend, lnox_cg_tend            & ! tendency (ppmv/s)
+!              )
+      CALL lightning_nox_decaria_cup ( &
               ! Frequently used prognostics
                 dx, dy, xland, ht, t_phy, rho, z, p_phy,      &
                 ic_flashrate, cg_flashrate,           & ! flashes (#/s)
               ! Scheme specific prognostics
+                cldfra_cup, htop,                     &
                 refl,                                 &
               ! Namelist inputs
                 N_IC, N_CG,                           &
@@ -141,6 +174,7 @@
               ! outputs
                 lnox_ic_tend, lnox_cg_tend            & ! tendency (ppmv/s)
               )
+!lm
     ! Invalid lightning NOx options
     CASE DEFAULT
         WRITE(wrf_err_message, * ) ' lightning_nox_driver: The lightning nox option does not exist: lnox_opt = ', lnox_opt
-- 
2.31.1

